iFixit/alertinator
# Alertinator: Prevent multiples of the same alert

From @entendu 
--
_(previous PR was from `entendu:master` so I am making a new PR as a separate `ifixit/alertinator` branch)_
```
A few of my assumptions based off #6:

> Customizable intervals: Notify me when the check fails, when it's been failing for at least 30min, 6hours...

Since checks are (ostensibly) already on a timer, I think it makes more sense to notify after _X_ failures/_Y_ successes.

> Be applied per (check-function, alert level) tuple

The way that `check()` is implemented, this is a problem. The alert level is based on which Exception class is used (e.g., `AlertinatorWarningException`), and in the case of success, the exception is never thrown. I have some ideas for implementing a better exception architecture (which would fit naturally into #3). As it stands, my code will alert on check-function only.

> Resetting counters for a given check after a successful run (Or X successful runs)

Implemented

> Storing any state locally on the file-system in whatever format makes sense.

Implemented as an Interface so someone could extend it to, say, mysql instead of the filesystem.

> If $shouldAlert is true, the caller of this function will format a nice message using the $firstAlertTimestamp; something like "$alertMessage -- sustained for 30 minutes".

Somewhat annoying -- the message is defined when the Exception is thrown and as you know there is no `$e->setMessage()` method. I did some dirty business to help, but the messaging methods will have to be refactored to get the level of robustness you're looking for.

This PR should be fully backwards-compatible. Unit tests are included and docs are updated.
```

CR
--
It looks bad, but at least 20% of the changes are documentation-only. Go commit by commit, and notice any code differences where the authors change.

QA
--
- [ ] Check for an alert after X failures
  - [ ] Check that a fail clears the running count of successes
- [ ] Check for an all-clear message after Y successes
  - [ ] Check that all-clear clears the running count of fails
- [ ] Check for an alert after every `$fails % $remindEvery == 0`
  - [ ] Check that alert reminder/spam prevention resets at Y number of successes
- [ ] Test backwards compatibility
  - Could be done by loading this branch as a git module on cominor and triggering an alert/all-clear cycle
  - Could also be done by loading this branch and using a `$config` from prior versions

*Run this `AlertinatorQA.php` as a test driver for a local Alertinator clone. PM me for necessary credentials.*
```php
<?php

require 'Alertinator.php';


$tester = (new AlertinatorQA());
$tester->check();

$GLOBALS['pf'] = true;

function testChecks() {
   $pf = $GLOBALS['pf'] ? 'passes' : 'fails';
   echo("Test " . $pf . "\n");
   if($GLOBALS['pf'])
      return;
   else
      throw new AlertinatorNoticeException("{$pf}");
}


class AlertinatorQA {
   public function __construct() {
      $myNumber = readline("Enter 10 digit phone number with no spaces, dashes or parens: ");
      $email = readline("Enter email: ");
      $sid = readline("Enter sid: ");
      $token = readline("Enter token: ");
      $alertAfter = intval(readline("alert after X fails: "));
      $clearAfter = intval(readline("clear after X passes: "));
      $remindEvery = intval(readline("remind every X failures: "));
      $config = [
         'twilio' => [
            'fromNumber' => '+18058745977',
            'accountSid' => '$sid',
            'authToken' => '$token',
         ],
         'checks' => [
            'testChecks' => [
                'groups' => ['default'],
                'alertAfter' => $alertAfter,
                'clearAfter' => $clearAfter,
                'remindEvery' => $remindEvery,
            ],
         ],
         'groups' => [
            'default' => [
               'me',
            ],
         ],
         'alertees' => [
            'me' => [
               'email' => [$email, Alertinator::NOTICE | Alertinator::WARNING],
               'sms' => [$myNumber, Alertinator::WARNING | Alertinator::CRITICAL],
               'call' => [$myNumber, /*Alertinator::CRITICAL*/ 0],
            ],
         ],
      ];

      $this->alertinator = new Alertinator($config);
   }

   public function check() {
      while(true) {
         $GLOBALS['pf'] = readline("p/f (pass/fail next test?): ") == "p";
         if($GLOBALS['pf'])
            echo "Passing...";
         else
            echo "Failing...";
         $this->alertinator->check($GLOBALS['pf']);
      }
   }
}
```

Closes #6 
Closes on submodule update https://github.com/iFixit/ifixit/issues/22839
